File: producer.py
from kafka import KafkaProducer
import json
import time

# Kafka Producer
producer = KafkaProducer(

    # Kafka broker address
    bootstrap_servers="192.168.29.13:9092",

    # Wait for all replicas acknowledgment
    # Gives maximum durability
    acks="all",

    # Retry failed sends
    retries=5,

    # Wait 50ms to create batches
    # Improves throughput
    linger_ms=50,

    # Batch size in bytes
    batch_size=32768,

    # Prevent duplicate messages during retries
    enable_idempotence=True
)

i = 0

while True:

    i += 1

    # Example order payload
    order = {
        "order_id": i,
        "amount": 100,
        "status": "NEW",
        "timestamp": time.time()
    }

    # Convert dict -> JSON -> bytes
    data = json.dumps(order).encode()

    try:

        # Send message to Kafka topic
        future = producer.send("stress-topic", data)

        # IMPORTANT:
        # Wait for Kafka acknowledgment
        # Prevents memory overflow/backpressure issues
        future.get(timeout=10)

        print(f"PRODUCED {i}")

    except Exception as e:

        print("PRODUCER ERROR:", e)

        # Avoid retry storm
        time.sleep(2)

File: src/producer.ts
// ===============================
// PRODUCER
// ===============================

import { kafka } from "./config/kafka";

/**
 * Kafka Producer
 *
 * Producer pushes messages/events into Kafka topic.
 *
 * In real systems:
 * - order service
 * - payment service
 * - inventory service
 *
 * all produce events into Kafka.
 */
const producer = kafka.producer({

  /**
   * Prevent duplicate messages during retries.
   *
   * Example:
   * If network fails after sending,
   * producer may retry.
   *
   * Idempotent producer ensures
   * Kafka stores message only once.
   */
  idempotent: true,

  /**
   * Retry sending message if broker temporarily fails.
   *
   * Helps during:
   * - broker restart
   * - network glitch
   * - temporary timeout
   */
  retry: {
    retries: 5,
  },
});

async function runProducer() {

  /**
   * Connect producer to Kafka broker.
   */
  await producer.connect();

  let i = 0;

  /**
   * Infinite loop to continuously
   * generate events/messages.
   */
  while (true) {

    i++;

    /**
     * Example event payload.
     */
    const order = {
      order_id: i,
      amount: 100,
      status: "NEW",
      timestamp: Date.now(),
    };

    try {

      /**
       * Send message to Kafka topic.
       */
      await producer.send({

        /**
         * Kafka topic name.
         *
         * Topic stores stream of events.
         */
        topic: "stress-topic",

        /**
         * acks = -1
         *
         * Wait for ALL Kafka replicas
         * to acknowledge write.
         *
         * Gives maximum durability.
         */
        acks: -1,

        /**
         * Array of Kafka messages.
         */
        messages: [
          {

            /**
             * Kafka stores bytes internally.
             *
             * Convert object -> JSON string.
             */
            value: JSON.stringify(order),
          },
        ],
      });

      console.log(`PRODUCED ${i}`);

      /**
       * Small delay.
       *
       * Prevents flooding broker
       * with millions of requests instantly.
       */
      await sleep(1000);

    } catch (error) {

      /**
       * Producer send failure.
       *
       * Common reasons:
       * - broker unavailable
       * - timeout
       * - network issue
       */
      console.error("PRODUCER ERROR:", error);

      /**
       * Wait before retrying.
       *
       * Prevents retry storm.
       */
      await sleep(2000);
    }
  }
}

/**
 * Utility function to simulate delay.
 */
function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Start producer process.
 */
runProducer().catch(console.error);
